### DISCUSSION TITLE
5.8 JsonServiceClient - BaseGetAllKeys Null Reference on Client.Get / Client.Post

### DISCUSSION THREAD

**Discussion by zerokewl88** [ServiceStack/Discuss#56] [Q&A]
Hi there, 

We have been struggling with a intermittent issue recently, and was wondering if anyone here is able to shed some light on this situation please. We find our JsonServiceClient begins to throw exceptions and the StackTrace provided below into our error logging shows the following error. 

<img width="791" alt="image" src="https://user-images.githubusercontent.com/5489947/223370877-6ce360ec-6f8a-4071-a4a4-7e3cd0b7f4e6.png">

Our ServiceStack implementation is hosted alongside a MVC implementation, and works perfectly. However when this exception occurs - our API services go down, and don't restore themselves until after manual intervention and a restart of IIS / server. 

All of our JsonServiceClient requests - all throw the above errors when attempting to send any sort of request.

We are using ServiceStack 5.8 and use a BaseService Implementation for all our API Services that inherits from ServiceStack.Service class - and we inject our JsonServiceClient on this base class which all our API Services start to use.

What we also then find, when this issue presents itself, is that our other 'Non-API Specific ServiceStack JsonClientService requests still works allright as users are still able to login into our site, and navigate around which all utilizes ServiceStack JsonServiceClient - but on a different 'area' / set of controllers in MVC. 

```
public class ApiServiceBase : Service
{
  public ApiServiceBase(JsonServiceClient client)
  {

  }
}
```

and a example of 

```
public class OurService : ApiServiceBase
{
  public AddPersonService(JsonServiceClient client) : base(client) { }

  public object Post(MyRequest request) 
  {
    var response = client.get(MyRequestDTO)  <---------- this is the where it starts to throw the above errors. 
  }
}
```

Note: We are at the point now where we need this resolved, and if we need direct support by purchasing the Business Perpetual License from ServiceStack - we also don't mind doing at this point, if we can find a stable resolution and move on from this annoying issue.

Any information / advice would be greatly appreciated.

Thank you.

**Comment by mythz** [comment]
Looks like the Exception is happening from within the `JsonServiceClient` when [adding the configured HttpHeaders to the WebRequest](https://github.com/ServiceStack/ServiceStack/blob/main/ServiceStack/src/ServiceStack.Client/PclExportClient.cs#L147):

```csharp
public virtual void AddHeader(WebRequest webReq, NameValueCollection headers)
{
    if (headers == null)
        return;

    foreach (var name in headers.AllKeys)
    {
        webReq.Headers[name] = headers[name];
    }
}
```

So I'd be looking at making sure the `client.Headers` are not modified at runtime if this client instance is used across multiple threads. 

If the client isn't stateful, you can inject a client factory instead, e.g. injecting a factory function like `Func<JsonServiceClient>` so your Service creates and uses an new JsonServiceClient instance.

If your API can be made to be async you can try switching to use a [JsonHttpClient](https://docs.servicestack.net/csharp-client#jsonhttpclient) instead.

If you're trying to call an internal ServiceStack Service (i.e. within the same AppHost) you should change to use the [Service Gateway](https://docs.servicestack.net/service-gateway) instead.

That's about the only thing I can recommend without a stand-alone repro (using the latest version) that we can run locally to repro the issue.

**Reply by zerokewl88** [reply]
Hi Mythz, 

It's strange, but your explanation above also does make sense to me as that the problem is intermittent, and so multiehtreaded issues that could cause the headers to be updated / changed, may be the issue like you have described.

I wont know this until it goes out and further testing occurs - i'll post any updates after making changes if this has resolved the issue. 

Thank you.
